File: Objective-C/LRI/Drink MVP/Views/DrinkCell.m
//
//  DrinkCell.m
//  LRI
//
//  Created by Development on 05.07.2022.
//

#import "DrinkCell.h"
#import "UIImageView+Extension.h"

@interface DrinkCell ()
@property (strong, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (strong, nonatomic) IBOutlet UIImageView *previewImageView;
@property (strong, nonatomic) IBOutlet UILabel *descriptionLabel;
@end

@implementation DrinkCell

+ (NSString *) identifier {
    return @"DrinkCell";
}

- (void)awakeFromNib {
    [super awakeFromNib];
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];
}

- (void)display:(DrinkItem *) item {
    self.cachedUrl = item.strDrinkThumb;
    
    self.descriptionLabel.text = item.strDrink;
    
    [self.previewImageView asyncImageWith:item.strDrinkThumb
                          withPlaceholder:[UIImage imageNamed:@"placeholder"]
                           withCompletion:^(UIImage * _Nonnull image, BOOL wasCached) {
        __weak typeof(self) weakSelf = self;
        if (wasCached || self.cachedUrl == item.strDrinkThumb) {
            weakSelf.previewImageView.image = image;
            [weakSelf.activityIndicator stopAnimating];
        }
    }];
}

@end

File: SwiftUIKit/LRI/Drink MVP/Views/DrinkCell.swift
//
//  DrinkTableViewCell.swift
//  LRI
//
//  Created by Development on 03.07.2022.
//

import UIKit

class DrinkCell: UITableViewCell {
    
    static let identifier: String = "DrinkCell"
    
    @IBOutlet var previewImageView: UIImageView!
    @IBOutlet var descriptionLabel: UILabel!
    @IBOutlet var activityIndicator: UIActivityIndicatorView!
    
    var cachedUrl: String?
    
    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)
        // Configure the view for the selected state
    }

    // if the cell is reusable (has a reuse identifier), this is called just before the cell is returned from the table view method dequeueReusableCellWithIdentifier:. If you override, you MUST call super.
    override func prepareForReuse() {
        super.prepareForReuse()
        // Update activity indicator
        self.activityIndicator.startAnimating()
    }
    
}

extension DrinkCell {
    
    /// Decorator for displaying drink item cell
    func display(_ item: DrinkItem?) {
        guard let item = item else { return }
        
        defer { self.cachedUrl = item.strDrinkThumb }
        
        self.previewImageView.asyncImage(with: item.strDrinkThumb) { [weak self] image, isCached in
            guard let self = self else { return }
            if isCached || self.cachedUrl == item.strDrinkThumb {
                self.previewImageView.image = image
                self.activityIndicator.stopAnimating()
            }
        }
        
        self.descriptionLabel.text = item.strDrink
    }
    
}
